python_modules/dagster/dagster/core/asset_defs/assets.py

```diff
@@ -146,20 +160,43 @@ def asset_keys(self) -> AbstractSet[AssetKey]:
         return set(self.asset_keys_by_output_name.values())
 
     @property
-    def asset_keys_by_output_name(self) -> Mapping[str, AssetKey]:
+    def dependency_asset_keys(self) -> Iterable[AssetKey]:
+        return self.asset_keys_by_input_name.values()
+
+    @property
+    def optional_asset_keys(self) -> AbstractSet[AssetKey]:
+        """
+        The set of asset keys that the underlying computation can produce but is not required to
```

Comment:
Is the set of asset keys produced up to the computation or up to the selection? It would be helpful to clarify.

Reply:
basically, we can't guarantee that an op's computation will actually respect the subset request. In general, I don't think it's a bad thing if the op generates MORE assets than expected, as at worst this will just result in redundant / wasteful computation

Reply:
I think it can be a bad thing because rematerializing is normally overwriting data - if we don't error in this case I think we'd at least want to warn pretty loudly.

Reply:
added a warning when we detect an asset key that is not expected to be produced